### PR TITLE
Add WASI example

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Work in progress, points to the branch on wasmerio/wasmer; we should merge this example after we merge the wasmer pr

### CODE REVIEW

**Review by torch2424 — COMMENTED** [review]
So I was going to approve this, but this is some heavy C for me 😂 Everything else looks fine though! Let me know if you want me to approve anyways! 👍

**Comment by MarkMcCaskey** [comment]
@torch2424 hahah, thanks!  It's actually mostly copied and pasted from the other example.  The total lines of new C code in this PR is ~20-40.  This PR is blocked on shipping the change onto Wasmer master so we don't point the cmake file to my branch.  Shipping it on Wasmer master is blocked by getting a working implementation in another language integration, which I'm working on now and then I'll ship these!

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
+add_executable(wasmer-wasi-c-api-example wasmer-wasi-c-api-example.c)
 
 include(ExternalProject)
 set_directory_properties(PROPERTIES EP_PREFIX ${CMAKE_BINARY_DIR}/rust-build)
 ExternalProject_Add(
         wasmer-runtime-c-api
         DOWNLOAD_DIR ${CMAKE_CURRENT_SOURCE_DIR}/build
         GIT_REPOSITORY https://github.com/wasmerio/wasmer.git
-        GIT_TAG master
+        GIT_TAG origin/feature/wasi-in-c-api
```
Comment (Hywan) [inline CMakeLists.txt]: Must not be merged as is.

File: wasm-wasi-sample-app/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# WASI Sample App
+
+## Requirements
+- describe WASI clang installation process here
```
Comment (Hywan) [inline wasm-wasi-sample-app/README.md]: Please, do :-p.

File: wasmer-wasi-c-api-example.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            print_byte_array(&import.import_name);
+            putchar('\n');
+
+            assert(import.tag == WASM_FUNCTION);
+            assert(import.value.func);
+            wasmer_import_object_imports_destroy(&import, 1);
+    }
+    wasmer_import_object_iter_destroy(func_iter);
+    func_iter = NULL;
+
```
Comment (Hywan) [inline wasmer-wasi-c-api-example.c:142]: Keep one empty line here.

File: wasmer-wasi-c-api-example.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    wasmer_import_object_t *import_object =
+            wasmer_wasi_generate_import_object(args, num_args, envs, num_envs, NULL, 0, NULL, 0);
+    // Define an array containing our imports
+    wasmer_import_t extra_imports[] = {func_import, global_import, memory_import};
+
+    // Add our custom imports to the WASI ImportObject
+    wasmer_import_object_extend(import_object, extra_imports, 3);
+
+    // Read the wasm file bytes
+    FILE *file = fopen("wasm-wasi-sample-app/wasi_example.wasm", "r");
```
Comment (Hywan) [inline wasmer-wasi-c-api-example.c:114]: I would rename `source.c` to `program.c`, and `wasi_example.wasm` to `program.wasm`, thoughts?

**Review by Hywan — COMMENTED** [review]
I think it would be a good idea to showcase wasienv (https://github.com/wasienv/wasienv), what do you think?

**Pull request merged**